api/src/main/java/org/openmrs/module/fhir2/api/dao/impl/FhirServiceRequestDaoImpl.java

```diff
@@ -84,4 +103,70 @@ private void handleCodedConcept(Criteria criteria, TokenAndListParam code) {
 		        handleDate("autoExpireDate", dateRangeParam.getUpperBound())))))))));
 	}
 	
+	private void handleHasAndParam(Criteria criteria, HasAndListParam hasAndListParam) {
+		if (hasAndListParam == null) {
+			return;
+		}
+		
+		HasParam param = hasAndListParam.getValuesAsQueryTokens().get(0).getValuesAsQueryTokens().get(0);
```

Comment:
So, there's a quite a bit of refactoring that should be done around this. The reason there are so many handlers is really to try to simplify the code. So something like this might work better (this replaces basically `handleHasAndParam`, `handleHasOrParam`, and `handleHasParam`)

```java
	private void handleHasAndParam(Criteria criteria, HasAndListParam hasAndListParam) {
		if (hasAndListParam == null || hasAndListParam.size() == 0) {
			return;
		}

		DetachedCriteria obsSubquery = DetachedCriteria.forClass(Obs.class, "obs");
		handleAndListParam(hasAndListParam, hasParam -> handleHasParam(obsSubquery, hasParam)).ifPresent(obsSubquery::add);
		obsSubquery.setProjection(property("obs.order.orderId"));
		criteria.add(exists(obsSubquery));
	}

	private Optional<Criterion> handleHasParam(DetachedCriteria criteria, HasParam hasParam) {
		switch (hasParam.getParameterName()) {
			case Observation.SP_ENCOUNTER:
				if (lacksAlias(criteria, "e")) {
					criteria.createAlias("encounter", "e");
				}
				return Optional.of(eq("e.uuid", hasParam.getParameterValue()));
			break;
			case Observation.SP_PATIENT:
				// etc;
		}
	}
```

Note that this implementation relies on a couple of functions I added to this PR in `BaseDao`. This is intended to be something of a pattern for how things are handled, made just a bit more complicated by the need to do this for a subquery.